File: src/java/main/src/androidTest/java/com/tencent/wcdb/winq/ColumnDefTest.java
// Created by qiuwenchen on 2023/4/7.
//

/*
 * Tencent is pleased to support the open source community by making
 * WCDB available.
 *
 * Copyright (C) 2017 THL A29 Limited, a Tencent company.
 * All rights reserved.
 *
 * Licensed under the BSD 3-Clause License (the "License"); you may not use
 * this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 *       https://opensource.org/licenses/BSD-3-Clause
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.tencent.wcdb.winq;

import static com.tencent.wcdb.base.TestTool.winqEqual;

import androidx.test.ext.junit.runners.AndroidJUnit4;

import org.junit.Test;
import org.junit.runner.RunWith;

@RunWith(AndroidJUnit4.class)
public class ColumnDefTest {
    @Test
    public void test() {
        winqEqual(new ColumnDef("testColumn"), "testColumn");
        winqEqual(new ColumnDef("testColumn", ColumnType.Integer), "testColumn INTEGER");
        winqEqual(new ColumnDef(new Column("testColumn")), "testColumn");
        winqEqual(new ColumnDef(new Column("testColumn"), ColumnType.Integer), "testColumn INTEGER");
        winqEqual(new ColumnDef("testColumn", ColumnType.Integer), "testColumn INTEGER");
        winqEqual(genColumnDef().makePrimary(), "testColumn INTEGER PRIMARY KEY");
        winqEqual(genColumnDef().makePrimary(true), "testColumn INTEGER PRIMARY KEY AUTOINCREMENT");
        winqEqual(genColumnDef().makeNotNull(), "testColumn INTEGER NOT NULL");
        winqEqual(genColumnDef().makeDefaultTo(LiteralValue.currentDate()), "testColumn INTEGER DEFAULT CURRENT_DATE");
        winqEqual(genColumnDef().makeDefaultTo(LiteralValue.currentTime()), "testColumn INTEGER DEFAULT CURRENT_TIME");
        winqEqual(genColumnDef().makeDefaultTo(LiteralValue.currentTimeStamp()), "testColumn INTEGER DEFAULT CURRENT_TIMESTAMP");
    }

    ColumnDef genColumnDef() {
        return new ColumnDef("testColumn", ColumnType.Integer);
    }
}

File: src/rust/examples/tests/winq/column_def_test.rs
#[cfg(test)]
pub mod column_def_test {
    use crate::base::winq_tool::WinqTool;
    use wcdb::winq::column::Column;
    use wcdb::winq::column_def::ColumnDef;
    use wcdb::winq::column_type::ColumnType;
    use wcdb::winq::foreign_key::ForeignKey;
    use wcdb::winq::literal_value::LiteralValue;

    #[test]
    pub fn test() {
        WinqTool::winq_equal(&ColumnDef::new("testColumn"), "testColumn");
        WinqTool::winq_equal(
            &ColumnDef::new(("testColumn", ColumnType::Integer)),
            "testColumn INTEGER",
        );
        WinqTool::winq_equal(
            &ColumnDef::new(&Column::new("testColumn", None)),
            "testColumn",
        );
        WinqTool::winq_equal(
            &ColumnDef::new((&Column::new("testColumn", None), ColumnType::Integer)),
            "testColumn INTEGER",
        );
        WinqTool::winq_equal(
            &ColumnDef::new(("testColumn", ColumnType::Integer)),
            "testColumn INTEGER",
        );
        WinqTool::winq_equal(
            gen_column_def().make_primary(None),
            "testColumn INTEGER PRIMARY KEY",
        );
        WinqTool::winq_equal(
            gen_column_def().make_primary(Some(true)),
            "testColumn INTEGER PRIMARY KEY AUTOINCREMENT",
        );
        WinqTool::winq_equal(
            gen_column_def().make_primary(Some(false)),
            "testColumn INTEGER PRIMARY KEY",
        );
        WinqTool::winq_equal(
            gen_column_def().make_not_null(),
            "testColumn INTEGER NOT NULL",
        );
        WinqTool::winq_equal(
            gen_column_def().make_default_to(&LiteralValue::current_date()),
            "testColumn INTEGER DEFAULT CURRENT_DATE",
        );
        WinqTool::winq_equal(
            gen_column_def().make_default_to(&LiteralValue::current_time()),
            "testColumn INTEGER DEFAULT CURRENT_TIME",
        );
        WinqTool::winq_equal(
            gen_column_def().make_default_to(&LiteralValue::current_time_stamp()),
            "testColumn INTEGER DEFAULT CURRENT_TIMESTAMP",
        );
        WinqTool::winq_equal(
            gen_column_def().make_foreign_key(&ForeignKey::new()),
            "testColumn INTEGER REFERENCES ",
        );
        WinqTool::winq_equal(gen_column_def().make_unique(), "testColumn INTEGER UNIQUE");
        WinqTool::winq_equal(
            gen_column_def().make_not_indexed(),
            "testColumn INTEGER UNINDEXED",
        );
    }

    fn gen_column_def() -> ColumnDef {
        ColumnDef::new(("testColumn", ColumnType::Integer))
    }
}
